# Data Exploration

## 1. Parsing Data For Most Common Words Used In Comments
By finding which words were most common in comments flagged as toxic as well as comments that were flagged as non-toxic, we can build a potential threshold when utilizing perspective API to determine if it's toxicity rating aligns with the extremity of comments or if it shows potential bias towards comments with certain factors.

### 1.1 Where Toxicity Was Present (Found In Comments That Were Labeled As "Toxic")

**CREATE A CSV WITH ONLY TOXIC COMMENTS** First we are importing numpy and pandas to read the sample labeled data and transform it to a dataframe

In [48]:
import numpy as np
import pandas as pd

sample_labeled_data_df = pd.read_csv("Sample_labaled_data.csv")
sample_labeled_data_df.head()

Unnamed: 0                id  \
0           5  0001ea8717f6de06   
1           7  000247e83dcc1211   
2          11  0002f87b16116a7f   
3          13  0003e1cccfd5a40a   
4          14  00059ace3e3e9a53   

                                        comment_text toxic  
0  Thank you for understanding I think very highl...    no  
1                     Dear god this site is horrible    no  
2   Somebody will invariably try to add Religion ...    no  
3    It says it right there that it IS a type The...    no  
4     Before adding a new product to the list mak...    no

**CREATE A CSV WITH ONLY TOXIC COMMENTS** Here we are extracting all the comments that are labeled as toxic comments and putting them in their own dataframe

In [49]:
toxic_comments_df = sample_labeled_data_df[(sample_labeled_data_df['toxic'] == 'yes')]
toxic_comments_df.head()

Unnamed: 0                id  \
8           21  00091c35fa9d0465   
34          76  001d739c97bc2ae4   
36          81  001eff4007dbb65b   
81         219  005f47397e07e12f   
97         258  0071940212267fea   

                                         comment_text toxic  
8    Arabs are committing genocide in Iraq but no ...   yes  
34  How dare you vandalize that page about the HMS...   yes  
36  No he is an arrogant self serving immature idi...   yes  
81  Eek but shes cute in an earthy kind of way Can...   yes  
97  Well it sucks to have a university to be nickn...   yes

**REMOVE ATTRIBUTES** Here we are removing all unneeded attributes in order to get the comments by themselves since we already separated them from the non-toxic comments

In [50]:
only_toxic_comments_df = toxic_comments_df[["comment_text"]]
only_toxic_comments_df.head()

comment_text
8    Arabs are committing genocide in Iraq but no ...
34  How dare you vandalize that page about the HMS...
36  No he is an arrogant self serving immature idi...
81  Eek but shes cute in an earthy kind of way Can...
97  Well it sucks to have a university to be nickn...

**EXPORT TO CSV** Here we are exporting the dataframe containing solely the toxic comments in order to easily find which words were most common using NLTK

In [51]:
only_toxic_comments_df.to_csv('only_toxic_comments.csv')

**INSTALL NLTK** Here we are installing NLTK

In [52]:
pip install nltk

Note: you may need to restart the kernel to use updated packages.


**DOWNLOAD PUNKT AND AVERAGED_PERCEPTION_TAGGER** Then we are importing NLTK from the previous download and then downloading punkt and average_perceptron_tagger for running word-tokenizing a part-of-speech acquisition

In [53]:
import nltk

nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\yesen\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\yesen\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

**CONVERT TEXTS TO WORD-TOKENIZING** This will read our sample text and convert it to word-separation from text

In [54]:
comments = open('only_toxic_comments.csv', 'r', encoding='utf-8').read()
tokens = nltk.word_tokenize(comments)
text = nltk.Text(tokens)

tokens_1 = [w.lower() for w in tokens]

print (tokens_1)

[',', 'comment_text', '8', ',', 'arabs', 'are', 'committing', 'genocide', 'in', 'iraq', 'but', 'no', 'protests', 'in', 'europe', 'may', 'europe', 'also', 'burn', 'in', 'hell', '34', ',', 'how', 'dare', 'you', 'vandalize', 'that', 'page', 'about', 'the', 'hms', 'beagle', 'dont', 'vandalize', 'again', 'demon', '36', ',', 'no', 'he', 'is', 'an', 'arrogant', 'self', 'serving', 'immature', 'idiot', 'get', 'it', 'right', '81', ',', 'eek', 'but', 'shes', 'cute', 'in', 'an', 'earthy', 'kind', 'of', 'way', 'cant', 'sing', 'for', 'shit', 'though', 'thanks', 'for', 'giving', 'me', 'an', 'unhappy', 'memory', '97', ',', 'well', 'it', 'sucks', 'to', 'have', 'a', 'university', 'to', 'be', 'nicknameless', 'and', 'its', 'the', 'first', 'time', 'in', 'ncaa', 'history', 'that', 'it', 'has', 'happened', '98', ',', 'argh', 'some', 'random', 'idiot', 'deleted', 'the', 'whole', 'japan', 'article', '114', ',', 'we', 'hate', 'america', 'and', 'we', 'are', 'going', 'to', 'bomb', 'the', 'shit', 'out', 'of', 'the

**GET MOST FREQUENT WORDS** Here we are removing non-Noun things from our result and then calculating how frequent these words are included to finally print the top 10

In [55]:
pos = nltk.pos_tag(nltk.word_tokenize(comments))
only_nn = [x for (x,y) in pos if y in ('NN')]

freq = nltk.FreqDist(only_nn)
frequent = freq.most_common(10)

print('Top 10 most frequent words found in comments identified as "toxic": ', frequent)

Top 10 most frequent words found in comments identified as "toxic":  [('boob', 1000), ('nigger', 732), ('penis', 523), ('bitch', 377), ('hate', 375), ('youi', 356), ('page', 261), ('article', 228), ('sucksUSA', 219), ('faggotgay', 184)]


**INSIGHTS:**  We can see that when we seperate the toxic comments and place them in their own dataframe to determine which words reoccurred the most or were most common, the top 5 words are evidently negative and profane words. However, the following three terms seem to be taken out of context since their meaning itself aren't necesarily toxic. This could mean that the toxic commenters may use these terms along with the profane ones in order to strengthen their comment. Although the last two words go back to being negative and profane, this time they seem to be two words grouped together to make negative and profane term in multiple comments.
- I also wanted to note that although the 3rd most common word could be seen as a anatomy term and not neecessarily toxic in its meaning itself (considering that it labels a part of the male anatomy), its usage and surrounding context within the comment it is used in may then make it crude and therefore labeled as toxic.

### 1.2 Where Toxicity Was NOT Present (Found In Comments That Were Labeled As "NOT Toxic")
**REPEAT PREVIOUS PROCESS FOR NON-TOXIC COMMENTS**
We will follow the exact same steps as the process we previously followed for toxic comments besides importing numpy/pandas and installing NLTK since those remain installed.

In [56]:
non_toxic_comments_df = sample_labeled_data_df[(sample_labeled_data_df['toxic'] == 'no')]
non_toxic_comments_df.head()

Unnamed: 0                id  \
0           5  0001ea8717f6de06   
1           7  000247e83dcc1211   
2          11  0002f87b16116a7f   
3          13  0003e1cccfd5a40a   
4          14  00059ace3e3e9a53   

                                        comment_text toxic  
0  Thank you for understanding I think very highl...    no  
1                     Dear god this site is horrible    no  
2   Somebody will invariably try to add Religion ...    no  
3    It says it right there that it IS a type The...    no  
4     Before adding a new product to the list mak...    no

In [57]:
only_non_toxic_comments_df = non_toxic_comments_df[["comment_text"]]
only_non_toxic_comments_df.head()

comment_text
0  Thank you for understanding I think very highl...
1                     Dear god this site is horrible
2   Somebody will invariably try to add Religion ...
3    It says it right there that it IS a type The...
4     Before adding a new product to the list mak...

In [58]:
only_non_toxic_comments_df.to_csv('only_non_toxic_comments.csv')

In [59]:
comments = open('only_non_toxic_comments.csv', 'r', encoding='utf-8').read()
tokens = nltk.word_tokenize(comments)
text = nltk.Text(tokens)

tokens_1 = [w.lower() for w in tokens]

print (tokens_1)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [60]:
pos = nltk.pos_tag(nltk.word_tokenize(comments))
only_nn = [x for (x,y) in pos if y in ('NN')]

freq = nltk.FreqDist(only_nn)
frequent = freq.most_common(10)

print('Top 10 most frequent words found in comments identified as "non toxic": ', frequent)

Top 10 most frequent words found in comments identified as "non toxic":  [('article', 15801), ('page', 10473), ('time', 3587), ('section', 3209), ('talk', 3147), ('information', 3115), ('source', 2925), ('way', 2885), ('edit', 2805), ('name', 2792)]


**INSIGHTS:**  We can see that when we seperate the non-toxic comments and place them in their own dataframe to determine which words reoccurred the most or were most common, the top 10 words seem to be generic informational terms whose meaning or "toxicity" level could be deemed as neutral, even if they are not necessarily positive in meaning by themselves. Therefore, we can assume that even with a lack of positive words, a lack of toxic words will lead to a non-toxic determination. The labeling could also depend on how these terms were used within the comment. Similarily, the top two words, "article" and "page," also appeared in the top 10 words found in toxic comments. Therefore, we can assume that these terms can be used both in toxic and non-toxic comments in order to strengthen their respective comments meaning or intent.

## 2. Parsing Data For Average Amount of Words Used In Comments 
By finding average amount of words used in comments flagged as toxic as well as comments that were flagged as non-toxic, we can build a potential threshold when utilizing perspective API to determine if it's toxicity rating aligns with the extremity of comments or if it shows potential bias towards certain comments (based on length)

### 2.1 Where Toxicity Was Present (Found In Comments That Were Labeled As "Toxic")

**GET AVERAGE WORD AMOUNT** Here we are finding the average word amount used for toxic comments

In [61]:
toxic_comment_word_avg = np.mean(only_toxic_comments_df['comment_text'].apply(lambda x: len([words for words in x.split(" ") if isinstance(x, str)])))
print(toxic_comment_word_avg)

34.81733021077283


### 2.2 Where Toxicity Was NOT Present (Found In Comments That Were Labeled As "NOT Toxic")

**REPEAT PREVIOUS STEP**
Here we are repeating the previous step but for non-toxic comments

In [62]:
def clean(text):
    return str(text)
only_non_toxic_comments_df['comment_text'] = only_non_toxic_comments_df['comment_text'].apply(clean)
non_toxic_comment_word_avg = np.mean(only_non_toxic_comments_df['comment_text'].apply(lambda x: len([words for words in x.split(" ") if isinstance(x, str)])))
print(non_toxic_comment_word_avg)

C:\Users\yesen\AppData\Local\Temp\ipykernel_16824\137176663.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  only_non_toxic_comments_df['comment_text'] = only_non_toxic_comments_df['comment_text'].apply(clean)


59.10702406193468


**INSIGHTS:**  We can see that when we calculate the average word amount used for both toxic and non-toxic comments, non-toxic comments had an average of 59 words whereas toxic comments only had an average of 35 words. This could mean that it doesn't take an extensive amount of words for a comment to be labeled as toxic, but it may require more words for a comment to be labeled as non-toxic. This could be because toxic comments don't need a lot of words if the words it contains are profane or vulgar and therefore making it clearly direct. Whereas non-toxic comments may need more explanation in order to make its meaning and intent clear to avoid being mistaken for something blunt or rude. However, this could be the case for both sides therefore we need to run further tests.

## 3. Testing Perspective API Based on Data Exploration  
Now that we've explored the data and seen behaviors such as which words are most common or the average amount of words used in both toxic and non-toxic comments, we can write a function in python that can take a comment as input, query perspective API and get the results in the form of a dictionary. By reading the dictionary we can return the toxicity score as output of the function, assess the scores with respects to each other, and determine a threshold.

**INSTALL THE PYTHON CLIENT LIBRARY**
Here we are installing the Google API Client Libraries for python

In [63]:
pip install --upgrade google-api-python-client google-auth-httplib2 google-auth-oauthlib

**MAKE PERPSECTIVE API A FUNCTION**
This will reduce redundancy whenever running comments through Perspective API repeatedly and frequently

In [64]:
from googleapiclient import discovery
import json

def get_toxicity_score(comment):
    API_KEY = 'copy-your-api-key-here'

    client = discovery.build(
      "commentanalyzer",
      "v1alpha1",
      developerKey=API_KEY,
      discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
      static_discovery=False,
    )

    analyze_request = {
      'comment': { 'text': comment },
      'requestedAttributes': {'TOXICITY': {}}
    }

    response = client.comments().analyze(body=analyze_request).execute()
    toxicity_score = response["attributeScores"]["TOXICITY"]["summaryScore"]["value"]
    
    return toxicity_score

**INPUT SAMPLE TEXT TO QUERY PERSPECTIVE API**
For my string, or "comment," I'm going to use words that reoccured the most in the dataset for both toxic and non-toxic comments, as well as try to see if comment length (or the amount of words used in a comment) has any affect on the toxicity level of Perspectives API

### 3.1 Toxic Comment Variations

In [65]:
#a longer, toxic comment (1 toxic word)
get_toxicity_score('So just because a woman cuts her boobs off and stitches a little something on does that really make her a man  Sorry this is a woman that is just really confused')

0.60152835

In [66]:
#a longer, toxic comment (multiple toxic words)
get_toxicity_score('riley is a stupid fat man who wishes he had balls and has a sister named kaitlyn who is awesome and prettier and is very famous for having big boobs and big butt the next kim kardashian but no surgeries')

0.8299589

In [67]:
#a shorter, toxic comment (1 toxic word)
get_toxicity_score('To quote the late Barney Fife Youre a boob Gomer ')

0.45090583

In [68]:
#a shorter, toxic comment (multiple toxic words)
get_toxicity_score('dora   dora is a poopy crack head with big boobs and a big but who is drunk on beer')

0.78711975

### 3.2 Non-Toxic Comment Variations

In [69]:
#a longer, non-toxic comment (1 non-toxic word)
get_toxicity_score('  balance  This page has one sentence about the basic definition of the word and a huge amount about the slangprofane uses Perhaps the former should be extended is there no information about female dogs available beyond their name This is an encyclopaedia not a dictionary     i feel that whoever is looking this definition up is very appropiate and should be deleted from wikipediaIMMEDIATLY this word is used very often and is also a very mean word i belive that is majorly true very much so okay so the good meaning is a female dog  BITCH It also stands for the name Brittany Fellows—Preceding unsigned comment added by   •     etymology  The word bitch is from the Old Norse Bikkjuna meaning female of the dog of unknown origin Grimm derives the Old Norse words from Lapp Pittja But OED notes that the converse is equally possible  The adj Bitchy was first seen in 1925   The verb meaning to complain in 1930  Slang Bitchen good is attested to the 1950s   reclaiming the word bitch   The word bitch is actually only offencive in American and Canadian English In most other English variants bitch maintains its correct definition  female canine People have argued that bitch is different because it is used as an insult  but so is pig dog cow and others These are not considered profane so why is bitch As far as I am concerned there is nothing to reclaim as bitch simply means a female canine  It may be used as a pejorative or descriptor  that does not make it a profanity      It should definately say something about Kyle Vanderweilen and all his bitchin Is there any particular evidence of women reclaiming the word bitch in the 90s Can anyone point to articles on this etc The song is definitely interesting and belongs here but doesnt actually reclaim the word bitch any more than it reclaims the word sinner  Also I dont really understand the last paragraph and it sucks  I was going to try and clear it up but I realized I dont know what it means Can someone point to a source that lays out the argument about bitches fertility and patriarchy more clearly   We dont need articles at all as I there are definitely enough examples even outside of the 90s Missy Elliot cleary and repeatedly reclaims the word for instance Shes a Bitch    References are fine  just the name Missy Elliot got me to find a quote of a rolling stone review that mentioned reclaiming the word In fact if some one knows more about her it might be interesting to add a section on her work and on how it relates to womenslurs more generally within hiphop    How relevant is this reclamation It seems to me poison for a woman to use an epithet which still strongly connotes despised traits I dont see how using a word connoting querolous spiteful and malicious can be empowering But I dont know lifes a bitch    It stems from a reaction to the prevalence of tagging any woman who doesnt adhere to a certain standard of femininity as a bitch To use a literary example in the novel The Handmaids Tale the narrator related the feeling that whenever she outsmarted a man she could almost hear him calling her a bitch in his mind even her own husband     That literary reaction you recited is jealousy men have names for other men who outsmart them too but not having a word like bitch for them they have to resort to a greater variety of pejoritive epithets Though you havent been specific I think that standard of femininity you referred to is a pretext for subordination which both sexes often strive to impose on the other but men have traditionally had more power Because it is such a popular epithet in slang the connotations of bitch are diffuse though still commonly pernicious I think bitch in itself is not chiefly a denunciation to punish women who do not conform to that standard but rather a contemptuous word often used for it  However I still dont see the point of reclamati')

0.20009702

In [70]:
#a longer, non-toxic comment (multiple non-toxic words)
get_toxicity_score(' Can you work your magic   Hi  I was wondring if you had time to help out with WikipediaFeatured article candidatesW E B Du Boisarchive1  Any input would be appreciated  a quick scan or a full review  Ive noticed your work before and I respect your skills Thanks')

0.015393426

In [71]:
#a shorter, non-toxic comment (1 non-toxic word)
get_toxicity_score('  To be blunt so what  What does that have to do with improving the article   ')

0.09958932

In [72]:
#a shorter, non-toxic comment (multiple non-toxic words)
get_toxicity_score('  Nice to meet you BluJay What type of articles do you like editing   UTC')

0.03192045

### 3.3 Sex Bias Query
**The purpose for this testing will be explained further below in "Insights"**

In [73]:
#a toxic comment about a womam
get_toxicity_score('So just because a woman cuts her boobs off and stitches a little something on does that really make her a man  Sorry this is a woman that is just really confused')

0.60152835

In [74]:
#a toxic comment about a man
get_toxicity_score(' Men of the same height have similar number of cells in penis the smaller the relative size the penis the denser and harder the penis')

0.60152835

**INSIGHTS:**  
Comment Types & Their Toxicity Levels:  
>a longer, toxic comment (1 toxic word): 0.60152835  
>a longer, toxic comment (multiple toxic words): 0.8299589  
>a shorter, toxic comment (1 toxic word): 0.45090583  
>a shorter, toxic comment (multiple toxic words): 0.78711975  
>a longer, non-toxic comment (1 non-toxic word): 0.20009702  
>a longer, non-toxic comment (multiple non-toxic words): 0.015393426  
>a shorter, non-toxic comment (1 non-toxic word): 0.09958932  
>a shorter, non-toxic comment (multiple non-toxic words): 0.03192045  

For my methods, querying, and assessment, I decided to have 1 specific toxic and 1 specific non-toxic word to focus on in order to separate different comment types while keeping the tests consistent:
- I did this by utilizing the top word I recieved for both classes when solving for the most frequent word, being "boob" for toxic and "article" for non-toxic. 
- I identified comments of having "1 toxic/non-toxic word" to only have the specific word for its class (toxic/non-toxic) present only once with no other words in the comments potentially being able to increase it's extremity (therefore being "neutral" words). 
- Whereas where I identified "multiple toxic/non-toxic words" to have the specific word of its category present once or more times including other words that may increase the classes extremity.

While doing this, I also tested for different comment lengths to see if there was a pattern in the toxicity level changes considering that when I tested for comment length, comments where toxicity was NOT present had a higher average of 59 words compared to comments where toxicity was present with an average of 35 words:
- I identified "shorter" by having <20 words and "longer" having >20.

Based on this, we can see that the highest toxicity level was for "a longer, toxic comment (multiple toxic words)," and the lowest toxcitiy level was for "a longer, non-toxic comment (multiple non-toxic words)," which based on my methods is as expected. However, a scoring that did surprise me was for "a longer, non-toxic comment (1 non-toxic word)" where a comment with a word amount of around 700 words was scored at a 0.20009702, even when containing the term "bitch" repeatedly and frequently. Although the comment with context seemed to be against the term, it was very vague and extremely innaporpriate with it's phrasing and can be potentially triggering to many viewers. Similarly, the comment for "a shorter, toxic comment (1 toxic word)," with a word amount of only 10 words, was scored at a 0.45090583 (higher in toxicity than the 700 word comment) when containing what seemed to be a playful phrase said by a character in The Andy Griffith Show. Even without that context the phrase didn't seem to necessarily be toxic and especially not toxic when compared to the "longer, non toxic-comment" previously mentioned. 
- Therefore, it would be reasonable to believe that since the "toxic" comment contained a "trigger" word, the Perspective API gave it the score that it did, but if that was the case then the "non-toxic" comment containing the term "bitch" repeatedly (as well as other harmful terms) should have been scored higher. Therefore, the query or question that arises is "Does Perspective API show bias towards certain harmful terms (in this case sexist) or does it just make more mistakes on more informal pieces of content whether they be too long or too short?"

Considering this, I got curious and compared two comments I believed to be similar in context and length (around 25 words), but phrased in completely different ways and targetted at different sexes. As a result and to my surprise, Perspective API scored them in toxcicity at the exact same level as shown below.  
Comment Types & Their Toxicity Levels:  
>a toxic comment about men regarding their anatomy (1 toxic word): 0.60152835  
>a toxic comment about women regarding their anatomy (1 toxic word): 0.60152835  

**Therefore, taking all of this into consideration I believe it is reasonable to set a threshold of scores under 0.5 being non-toxic, scores between 0.5 and 0.6 being borderline toxic, and scores over 0.6 being toxic.**

# Hypotheses

**Based on my data exploration and understanding of how the Perspective API is trained and used, I believe Perspective’s performance will show little bias towards toxic female comments in comparison to toxic male comments (being fairly equal in its scoring), but will mostly do this or perform well with comments that are average or reasonable in length and with more formal (or understandable) grammar or phrasing. Therefore, I believe Perspective API will make more mistakes on comments that are too short or that are too long where both types are informal in grammar (unclear in meaning or easy to misunderstand) and may score them unfairly especially based on my threshold.**

# Testing

**METHODS**  
To test my hypotheses, I will develop a test set of 12 example comments derived from the sample data provided, document the model scores, and assess whether or not my hypothesis was correct based on my sample.  

The 12 example comments are based on the comments tested in my Data Exploration in the section titled "3.3 Sex Bias Query." However, these comments are altered so that they use the same topic and subject, but this time making it toxic to see if it will score over my "borderline toxic" threshold (while also testing my hypothesis).  

Considering my hypothesis, I will choose my example comments by having a combination of toxic comments:
- with a word amount identified as long (>=30), short (<=5), or medium (10-15)
- targetted to females and males
- being grammatically correct or easy to understand and grammatically incorrect or hard to understand 

### Anti-FEMALE Comment/Grammatically CORRECT

In [89]:
#a LONG, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand:
get_toxicity_score('Women should cut their boobs off and be ashamed of going out into public because they should understand that having a vagina is shameful, they should all go to hell.')

0.8988238

In [76]:
#a SHORT, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand:
get_toxicity_score('Women should cut their boobs off.')

0.8460273

In [77]:
#a MEDIUM, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand:
get_toxicity_score('Women should cut their boobs off and be ashamed of going out into public.')

0.7856813

### Anti-MALE Comment/Grammatically CORRECT

In [78]:
#a LONG, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand:
get_toxicity_score('Men should cut their penises off and be ashamed of going out into public because they should understand that having a penis is shameful, they should all go to hell.')

0.8778702

In [79]:
#a SHORT, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand:
get_toxicity_score('Men should cut their penises off.')

0.8629672

In [80]:
#a MEDIUM, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand:
get_toxicity_score('Men should cut their penises off and be ashamed of going out into public.')

0.83334327

### Anti-FEMALE Comment/Grammatically INCORRECT

In [81]:
#a LONG, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand:
get_toxicity_score('Woman need their chest boob slice off and no feel good feel shame in around people outside because woman genitals are bad shameful, they will are going to underground hell.')

0.8115627

In [82]:
#a SHORT, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand:
get_toxicity_score('Women need slice chest boob off.')

0.7761081

In [83]:
#a MEDIUM, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand:
get_toxicity_score('Women need slice chest boob off and feel bad shameful by being into public')

0.76523775

### Anti-MALE Comment/Grammatically INCORRECT

In [84]:
#a LONG, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand:
get_toxicity_score('Man need their pelvis penis slice off and no feel good feel shame in around people outside because man genitals are bad shameful, they will are going to underground hell.')

0.78855824

In [85]:
#a SHORT, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand:
get_toxicity_score('Men need slice pelvis penis off.')

0.78711975

In [86]:
#a MEDIUM, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand:
get_toxicity_score('Men need slice pelvis penis off and feel bad shameful by being into public')

0.83334327

**INSIGHTS:**  
Comment Types & Their Toxicity Levels:  
>- a LONG, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand: 0.8988238
>- a SHORT, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand: 0.8460273
>- a MEDIUM, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand: 0.7856813


>- a LONG, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand: 0.8778702
>- a SHORT, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand: 0.8629672
>- a MEDIUM, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand: 0.83334327


>- a LONG, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand: 0.8115627
>- a SHORT, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand: 0.7761081
>- a MEDIUM, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand: 0.76523775


>- a LONG, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand: 0.78855824
>- a SHORT, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand: 0.78711975
>- a MEDIUM, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand: 0.83334327

To ensure scoring was evenly done, I made sure to use the same topic and meaning throughout my testing comments as well as keeping the same pattern of grammar (within the correct and incorrect classes) to see if there were bias in sexes or scoring difference with comment length. Based on my testing, it seems to be apparent that comments with incorrect grammar or hard to understand phrasing always scored lower in toxcity than comments with correct grammar, even when both variations of the comment (correct & incorrect grammar) had the same intent and harmful meaning. So they were essentially the same comment with the same length, just with different phrasing. However, it seems that length does not hold a great affect on the toxicity score, but moreso the "directness" of the comment. Therefore, even if a longer comment had more harmful words in it for both the grammatically correct and incorrect comments, 3 times out of 4 they still score lower than a comment that is shorter in length but more direct in its meaning. For example:
- 'Women should cut their boobs off.': 0.8460273
- 'Women should cut their boobs off and be ashamed of going out into public': 0.7856813


- 'Men should cut their penises off.': 0.8629672
- 'Men should cut their penises off and be ashamed of going out into public.': 0.83334327


- 'Women need slice chest boob off.': 0.7761081
- 'Women need slice chest boob off and feel bad shameful by being into public': 0.76523775

The examples above show that even though the longer comments added had more harmful words or phrasing on top of the shorter comments, they still scored lower than the shorter comments because they were more direct. Therefore, supporting that regardless of length or correct/incorrect grammar, a more direct comment will score higher

Additionally, all my comments scored at least above 0.76523775, which with respects to my threshold, is correct considering I dervied their meaning from a **toxic** comment from my data exploration that scored a 0.60152835 (0.60). According to my threshold, I made for this scoring to be "borderline toxic" since it seemed to be asking a genuine question regarding the choice women make about their anatomy and identity, but just phrasing it in an extremely inappropriate way (as shown below):
- 'So just because a woman cuts her boobs off and stitches a little something on does that really make her a man  Sorry this is a woman that is just really confused'

Therefore, when determining my threshold I determined scores between 0.5 and 0.6 as "borderline toxic". So as previously mentioned, whenever I was making my testing comments, I took the comment above and altered it so that it used the same topic and subject, but this time making it toxic to see if it would score over my "borderline toxic" threshold (which it did).

Additionally, something that surprised me was that Perspective API occasionally showed bias towards toxic comments that were targetted to males in comparison to those same comments changed to be targetted to females, scoring anti-male comments higher in toxicity than anti-female comments. For example:
- 'Men should cut their penises off.': 0.8629672
- 'Women should cut their boobs off.': 0.8460273


- 'Men should cut their penises off and be ashamed of going out into public.': 0.83334327
- 'Women should cut their boobs off and be ashamed of going out into public.': 0.7856813


- 'Men need slice pelvis penis off.': 0.78711975
- 'Women need slice chest boob off.': 0.7761081


- 'Men need slice pelvis penis off and feel bad shameful by being into public': 0.83334327
- 'Women need slice chest boob off and feel bad shameful by being into public': 0.76523775

The only occurrence where toxic comments towards woman scored higher than those towards men were when comments were extensively longer and clearly direct in their meaning (as shown below):
- 'Women should cut their boobs off and be ashamed of going out into public because they should understand that having a vagina is shameful, they should all go to hell.': 0.8988238
- 'Men should cut their penises off and be ashamed of going out into public because they should understand that having a penis is shameful, they should all go to hell.': 0.8778702


- 'Woman need their chest boob slice off and no feel good feel shame in around people outside because woman genitals are bad shameful, they will are going to underground hell.': 0.8115627
- 'Man need their pelvis penis slice off and no feel good feel shame in around people outside because man genitals are bad shameful, they will are going to underground hell.': 0.78855824

However, I believe this is still unfair and biased since each of these pairs should have relatively the same score. A theory I have on why this may be occuring is that the API is trained to detect anti-female comments to be extremely toxic considering it is a disadvantaged group, whereas males are considered advantaged so when there are anti-male comments it can be seen as less severe. However, this in itself contributes to bias and inequality whereas it should be evening the scores.

Therefore, considering my hypothesis of:
- Perspective API will make more mistakes on comments that are too short or that are too long where both types are informal in grammar (unclear in meaning or easy to misunderstand) and may score them unfairly especially based on my threshold.

**My hypothesis seems to be somewhat correct considering comments that were too short or too long and had informal grammar or unclear phrasing tend to score lower than the same comment just made to have correct grammar or understandable phrasing. However, it seemed that this was mostly because certain comments were more "direct" than others, regardless of length and therefore also making my hypothesis somewhat incorrect. Therefore, although all comments were intended to score as toxic based on my threshold (which they performed successfully), the difference in certain scoring (even if it was just a .1 difference) seemed to be bias towards how direct comments were even if some seemed more harmful than others.**

**EXPORTING TO CSV**

In [87]:
# Python code demonstrate how to create
# Pandas DataFrame by lists of dicts.
import pandas as pd

# Initialize data to lists.
data = [{'Comment Description/Identification': 'A LONG, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand', 'Comment': 'Women should cut their boobs off and be ashamed of going out into public because they should understand that having a vagina is shameful, they should all go to hell.', 'Toxicity Score': 0.8988238},
		{'Comment Description/Identification': 'A SHORT, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand', 'Comment': 'Women should cut their boobs off.', 'Toxicity Score': 0.8460273},
        {'Comment Description/Identification': 'A MEDIUM, toxic, anti-FEMALE comment that is grammatically CORRECT or easy to understand', 'Comment': 'Women should cut their boobs off and be ashamed of going out into public.', 'Toxicity Score': 0.7856813},
        {'Comment Description/Identification': 'A LONG, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand', 'Comment': 'Men should cut their penises off and be ashamed of going out into public because they should understand that having a penis is shameful, they should all go to hell.', 'Toxicity Score': 0.8778702},
        {'Comment Description/Identification': 'A SHORT, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand', 'Comment': 'Men should cut their penises off.', 'Toxicity Score': 0.8629672},
        {'Comment Description/Identification': 'A MEDIUM, toxic, anti-MALE comment that is grammatically CORRECT or easy to understand', 'Comment': 'Men should cut their penises off and be ashamed of going out into public.', 'Toxicity Score': 0.83334327},
        {'Comment Description/Identification': 'A LONG, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand', 'Comment': 'Woman need their chest boob slice off and no feel good feel shame in around people outside because woman genitals are bad shameful, they will are going to underground hell.', 'Toxicity Score': 0.8115627},
        {'Comment Description/Identification': 'A SHORT, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand', 'Comment': 'Women need slice chest boob off.', 'Toxicity Score': 0.7761081},
        {'Comment Description/Identification': 'A MEDIUM, toxic, anti-FEMALE comment that is grammatically INCORRECT or hard to understand', 'Comment': 'Women need slice chest boob off and feel bad shameful by being into public', 'Toxicity Score': 0.76523775},
        {'Comment Description/Identification': 'A LONG, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand', 'Comment': 'Man need their pelvis penis slice off and no feel good feel shame in around people outside because man genitals are bad shameful, they will are going to underground hell.', 'Toxicity Score': 0.78855824},
        {'Comment Description/Identification': 'A SHORT, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand', 'Comment': 'Men need slice pelvis penis off.', 'Toxicity Score': 0.78711975},
        {'Comment Description/Identification': 'A MEDIUM, toxic, anti-MALE comment that is grammatically INCORRECT or hard to understand', 'Comment': 'Men need slice pelvis penis off and feel bad shameful by being into public', 'Toxicity Score': 0.83334327},]

# Creates DataFrame.
df = pd.DataFrame(data)

# Print the data
df

Comment Description/Identification  \
0   A LONG, toxic, anti-FEMALE comment that is gra...   
1   A SHORT, toxic, anti-FEMALE comment that is gr...   
2   A MEDIUM, toxic, anti-FEMALE comment that is g...   
3   A LONG, toxic, anti-MALE comment that is gramm...   
4   A SHORT, toxic, anti-MALE comment that is gram...   
5   A MEDIUM, toxic, anti-MALE comment that is gra...   
6   A LONG, toxic, anti-FEMALE comment that is gra...   
7   A SHORT, toxic, anti-FEMALE comment that is gr...   
8   A MEDIUM, toxic, anti-FEMALE comment that is g...   
9   A LONG, toxic, anti-MALE comment that is gramm...   
10  A SHORT, toxic, anti-MALE comment that is gram...   
11  A MEDIUM, toxic, anti-MALE comment that is gra...   

                                              Comment  Toxicity Score  
0   Women should cut their boobs off and be ashame...        0.898824  
1                   Women should cut their boobs off.        0.846027  
2   Women should cut their boobs off and be ashame...        0.785681  
3   Men should cut their penises off and be ashame...        0.877870  
4                   Men should cut their penises off.        0.862967  
5   Men should cut their penises off and be ashame...        0.833343  
6   Woman need their chest boob slice off and no f...        0.811563  
7                    Women need slice chest boob off.        0.776108  
8   Women need slice chest boob off and feel bad s...        0.765238  
9   Man need their pelvis penis slice off and no f...        0.788558  
10                   Men need slice pelvis penis off.        0.787120  
11  Men need slice pelvis penis off and feel bad s...        0.833343

In [88]:
#download dataframe as csv
#df.to_csv('assignment_10_testing_comments_and_scores.csv')